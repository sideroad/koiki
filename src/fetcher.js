
const exec = ({
    dispatch,
    client,
    url,
    method,
    values,
    start,
    success,
    fail,
    mode,
    credentials,
    headers,
    after = (_values, res) => new Promise(resolve => resolve(res))
  }) => {
  dispatch(start(fetch.values));
  return client
          .fetchJSON({
            url,
            method,
            values,
            mode,
            credentials,
            headers
          })
          .then(
            (res) => {
              return after(values, res).then(
                converted => {
                  dispatch(success({
                    body: converted,
                    res
                  }));
                  return Promise.resolve({
                    body: converted,
                    res: res.res
                  });
                }
              );
            },
            (res) => {
              dispatch(fail(res));
              return Promise.reject({
                body: res.body,
                res: res.res
              });
            }
          );
};

export default class fetcher {
  constructor({urls, dispatch, client}) {

    Object.keys(urls).map(resource => {
      this[resource] = {};

      Object.keys(urls[resource]).map(action => {
        this[resource][action] = (values, options = {}) => {
          const _values = Object.assign( {}, urls[resource][action].defaults || {}, values);
          return exec({
            dispatch,
            client,
            url: urls[resource][action].url,
            method: urls[resource][action].method,
            after: urls[resource][action].after,
            mode: options.mode !== undefined ? options.mode : urls[resource][action].mode,
            credentials: options.credentials !== undefined ? options.credentials : urls[resource][action].credentials,
            headers: options.headers !== undefined ? options.headers : urls[resource][action].headers,
            values: _values,
            start: () => ({
              values: _values,
              type: resource + '/' + action.toUpperCase() + '_START'
            }),
            success: ({ body, res }) => ({
              values: _values,
              body,
              res,
              type: resource + '/' + action.toUpperCase() + '_SUCCESS'
            }),
            fail: ({ body, res, err }) => ({
              values: _values,
              body,
              res,
              err,
              type: resource + '/' + action.toUpperCase() + '_FAIL'
            })
          });
        };
      });
    });
  }
}
