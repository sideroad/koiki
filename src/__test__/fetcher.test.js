import sinon from 'sinon';

import ApiClient from '../apiclient';
import Fetcher from '../fetcher';

import './fetcher.mock';

describe('client', () => {
  it('should be create app for client', () => {
    const dispatch = sinon.spy();
    const logger = sinon.spy();
    const client = new ApiClient({}, logger);
    const fetcher = new Fetcher({
      urls: {
        person: {
          gets: {
            method: 'GET',
            url: 'https://chaus.herokuapp.com/apis/koiki/people',
            after: (values, res) => {
              expect(values).toMatchSnapshot();
              expect(res.body).toMatchSnapshot();
              return Promise.resolve(res.body);
            },
            next: action => action.next ? `https://chaus.herokuapp.com${action.next}` : ''
          }
        },
        fruit: {
          gets: {
            method: 'GET',
            url: 'https://chaus.herokuapp.com/apis/koiki/fruits',
            cache: {
              server: true
            }
          }
        }
      },
      dispatch,
      client,
      type: 'server'
    });
    return fetcher.person.gets({
      offset: 0,
      limit: 5
    }).then(
      (res) => {
        expect(res.body).toMatchSnapshot();
        expect(res.res.ok).toEqual(true);
        return fetcher.person.gets.next()
          .then(
            () => fetcher.person.gets.next()
          )
          .then(
            () => fetcher.person.gets.next()
          )
          .then(
            () => {
              expect(dispatch.getCall(0).args[0].type).toBe('person/GETS_START');
              expect(dispatch.getCall(0).args[0].body).toBe(undefined);
              expect(dispatch.getCall(1).args[0].type).toBe('person/GETS_SUCCESS');
              expect(dispatch.getCall(1).args[0].body.prev).toBe(null);
              expect(dispatch.getCall(1).args[0].body.next).toBe('/apis/koiki/people?offset=5&limit=5');
              expect(dispatch.getCall(2).args[0].type).toBe('person/GETS_NEXT_START');
              expect(dispatch.getCall(2).args[0].body).toBe(undefined);
              expect(dispatch.getCall(3).args[0].type).toBe('person/GETS_NEXT_SUCCESS');
              expect(dispatch.getCall(3).args[0].body.prev).toBe('/apis/koiki/people?offset=0&limit=5');
              expect(dispatch.getCall(3).args[0].body.next).toBe('/apis/koiki/people?offset=10&limit=5');
              expect(dispatch.getCall(3).args[0].hasNext).toBe(true);
              expect(dispatch.getCall(4).args[0].type).toBe('person/GETS_NEXT_START');
              expect(dispatch.getCall(4).args[0].body).toBe(undefined);
              expect(dispatch.getCall(5).args[0].type).toBe('person/GETS_NEXT_SUCCESS');
              expect(dispatch.getCall(5).args[0].body.prev).toBe('/apis/koiki/people?offset=5&limit=5');
              expect(dispatch.getCall(5).args[0].body.next).toBe(null);
              expect(dispatch.getCall(5).args[0].hasNext).toBe(false);
            }
          )
          .then(
            () => fetcher.fruit.gets({
              offset: 0,
              limit: 5
            })
          )
          .then(
            () => fetcher.fruit.gets({
              offset: 0,
              limit: 5
            })
          )
          .then(() => {
            expect(logger.getCall(3).args[0]).toBe('## fetch ');
            expect(logger.getCall(3).args[1]).toBe('https://chaus.herokuapp.com/apis/koiki/fruits?offset=0&limit=5');
            expect(logger.getCall(4).args[0]).toBe('## return from cache ');
            expect(logger.getCall(4).args[1]).toBe('https://chaus.herokuapp.com/apis/koiki/fruits?offset=0&limit=5');
            expect(dispatch.getCall(6).args[0].type).toBe('fruit/GETS_START');
            expect(dispatch.getCall(7).args[0].type).toBe('fruit/GETS_SUCCESS');
            expect(dispatch.getCall(7).args[0].body.items.length).toBe(1);
            expect(dispatch.getCall(8).args[0].type).toBe('fruit/GETS_START');
            expect(dispatch.getCall(9).args[0].type).toBe('fruit/GETS_SUCCESS');
            expect(dispatch.getCall(9).args[0].body.items.length).toBe(1);
          });
      }
    );
  });
});
