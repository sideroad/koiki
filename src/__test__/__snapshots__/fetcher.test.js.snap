exports[`client should be create app for client 1`] = `
Object {
  "first": "/apis/koiki/people?offset=0&limit=25",
  "items": Array [
    Object {
      "age": 30,
      "createdAt": "2016-10-15T04:42:29.000Z",
      "hobbies": Object {
        "href": "/apis/koiki/people/sideroad/hobbies",
      },
      "id": "sideroad",
      "name": "sideroad",
      "updatedAt": "2016-10-15T04:42:29.000Z",
    },
    Object {
      "age": 30,
      "createdAt": "2016-10-15T04:42:48.000Z",
      "hobbies": Object {
        "href": "/apis/koiki/people/koiki/hobbies",
      },
      "id": "koiki",
      "name": "koiki",
      "updatedAt": "2016-10-15T04:42:48.000Z",
    },
  ],
  "last": "/apis/koiki/people?offset=0&limit=25",
  "limit": 25,
  "next": null,
  "offset": 0,
  "prev": null,
  "size": 2,
}
`;
